src/S3Writer.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\S3Writer;
+
+use Aws\S3\Exception\S3Exception;
+use Aws\S3\S3Client;
+use Aws\S3\S3MultiRegionClient;
+use GuzzleHttp\Exception\ClientException;
+use Keboola\Component\UserException;
+use Monolog\Logger;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Finder\SplFileInfo;
+
+class S3Writer
+{
+    /**
+     * @var Config
+     */
+    private $config;
+
+    /**
+     * @var Logger
+     */
+    private $logger;
+
+    /**
+     * Application constructor.
+     */
+    public function __construct(Config $config, Logger $logger)
+    {
+        $this->config = $config;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Runs data extraction
+     * @throws \Exception
+     */
+    public function execute(string $sourcePath): void
+    {
+        try {
+            $client = new S3MultiRegionClient([
+                'version' => '2006-03-01',
+                'credentials' => [
+                    'key' => $this->config->getAccessKeyId(),
+                    'secret' => $this->config->getSecretAccessKey(),
+                ],
+            ]);
+            $region = $client->getBucketLocation(["Bucket" => $this->config->getBucket()])->get('LocationConstraint');
+            $client = new S3Client([
+                'region' => $region,
+                'version' => '2006-03-01',
+                'credentials' => [
+                    'key' => $this->config->getAccessKeyId(),
+                    'secret' => $this->config->getSecretAccessKey(),
+                ],
+            ]);
+
+            $finder = (new Finder())->in($sourcePath)->files();
+            /** @var SplFileInfo $file */
+            foreach ($finder as $file) {
+                // Remove initial forwardslash
+                $prefix = $this->config->getPrefix();
+                if (substr($prefix, 0, 1) == '/') {
+                    $prefix = substr($prefix, 1);
+                }
+                $key = $prefix . $file->getRelativePathname();
+                $this->logger->info("Uploading file {$file->getRelativePathname()} to {$key}");
+                $client->putObject(
+                    [
+                        'Bucket' => $this->config->getBucket(),
+                        'Key' => $key,
+                        'Body' => fopen($file->getPathname(), 'r'),
+                    ]
+                );
+            }
+        } catch (S3Exception $e) {
```

Comment:
trochu bych to provzdušnil :) (=prázdné řádky)

Plus do komentářů, co která větev řeší pls...

Reply:
to není poznat z `$e->getStatusCode() === 40#`?

Reply:
Jo, je. Asi jsem měl nějakou postgastrální demenci :) Nějak mě tam zmátla ta logika s ClientException (btw, proč ne instance of?)

Reply:
čistě ze zvyku, stejně tak by tam mohlo bejt `in_array($e->getStatusCode(), [400, 401, 404])`

Reply:
No, ale ono to není to samé. 

* get_class() ti vrátí stringem název classy, který porovnáš proti jinému stringu (tj. "je tohle přesně přímo tahle exception class?")
* instanceof ti vrátí, jestli "můžeš tenhle objekt považovat a nakládat s ním jako s ConnectException". Takže třeba kdyby ti to vrátilo `DnsError extends ConnectException`, tak to správně projde.

Reply:
hotovo